api/tests/unit/infrastructure/repositories/certification-partner-repository_test.js

```diff
@@ -0,0 +1,75 @@
+const { expect, sinon, } = require('../../../test-helper');
+const certificationPartnerRepository = require('../../../../lib/infrastructure/repositories/certification-partner-repository');
+const competenceMarkRepository = require('../../../../lib/infrastructure/repositories/competence-mark-repository');
+const competenceRepository = require('../../../../lib/infrastructure/repositories/competence-repository');
+const badgeAcquisitionRepository = require('../../../../lib/infrastructure/repositories/badge-acquisition-repository');
+const CertificationPartnerBookshelf = require('../../../../lib/infrastructure/data/certification-partner');
+const Badge = require('../../../../lib/domain/models/Badge');
+const CertificationPartnerClea = require('../../../../lib/domain/models/CertificationPartnerClea');
+
+describe('Unit | Repository | Certification Partner', function() {
+  const certificationCourseId = Symbol('certifCourseId');
+  const userId = Symbol('userId');
+  const totalPixCleaByCompetence = Symbol('totalPixCleaByCompetence');
+  const badgeKey = Symbol('badgeKey');
+  const competenceMarks = Symbol('competenceMarks');
+  const hasAcquiredBadgeClea = Symbol('hasAcquired');
+  const reproducibilityRate = Symbol('reproducibilityRate');
+  const domainTransaction = Symbol('domainTransaction');
+
+  beforeEach(() => {
+    sinon.stub(competenceMarkRepository, 'getLatestByCertificationCourseId')
+      .withArgs({ certificationCourseId, domainTransaction })
+      .resolves(competenceMarks);
+    sinon.stub(competenceRepository, 'getTotalPixCleaByCompetence').withArgs().resolves(totalPixCleaByCompetence);
+    sinon.stub(badgeAcquisitionRepository, 'hasAcquiredBadgeWithKey').withArgs({
+      badgeKey: Badge.keys.PIX_EMPLOI_CLEA,
+      userId,
+    }).resolves(hasAcquiredBadgeClea);
+    sinon.stub(CertificationPartnerClea, 'create');
+  });
+
+  describe('#buildCertificationPartnerClea', () => {
+    it('should create a certification clea object', async () => {
+      const certificationPartner = await certificationPartnerRepository.buildCertificationPartnerClea({
+        certificationCourseId,
+        userId,
+        reproducibilityRate,
+        domainTransaction
+      });
+      expect(certificationPartner).to.be.not.null;
+      expect(competenceMarkRepository.getLatestByCertificationCourseId).to.have.been.called;
+      expect(badgeAcquisitionRepository.hasAcquiredBadgeWithKey).to.have.been.called;
+      expect(competenceRepository.getTotalPixCleaByCompetence).to.have.been.called;
+      expect(CertificationPartnerClea.create).to.have.been.calledWith({
+        certificationCourseId,
+        hasAcquiredBadgeClea,
+        competenceMarks,
+        totalPixCleaByCompetence,
+        reproducibilityRate
+      });
```

Comment:
Ici je crois qu'il n'y a pas d'indirect output (valeur de retour indirecte), est-ce qu'on peut tester la valeur retournée au lieu de vérifier que des méthodes ont été appelées ? 
Vérifier qu'une méthode a été appelée sert à vérifier une valeur de retour indirecte en général - sauvegarde en base, envoi de mail, ...

Reply:
Alors, non, parce que dans ce cas on testerait soit des valeurs privées, soit on test toute l'implementation ce qui n'est pas non plus désiré.
On a decidé de verifié les argument via JOI, le test sera fait dans l'implem, on pourra donc retirer cette partie

Reply:
👍, les `to.have.been.called` sur des méthodes `get…` ou `has…` sont suspects de manière générale parce qu'on n'a pas vraiment besoin de vérifier que ces méthodes ont été appelées, mais plutôt que les bonnes informations ont été utilisées pour construire la valeur retournée par la fonction. Dans ces cas-là je préfère utiliser des _stubs_ précis sur les `get…` et vérifier en sortie qu'on retrouve les valeurs injectées.
Ex. stub de `competenceRepository.getTotalPixCleaByCompetence` qui renvoie `123` puis vérifier que dans l'objet retourné on a bien `123` (ce qui vérifie au passage que `getTotalPixCleaByCompetence` a forcément été appelé, mais en plus qu'on a bien utilisé sa valeur).

Reply:
On avait au depart pas le create et on ne peux pas tester le constructeur. Donc on a utiliser les to.have.been.called.
Effectivement, maintenant qu'on test les parametres du create, plus besoin de verifier que les autres fonctions sont appelés

Reply:
Je suppose que tu veux dire qu'on ne peut pas intercepter les appels à un constructeur. Le fait qu'on puisse le faire avec une _factory method_ comme `create` n'en fait pas automatiquement une bonne idée 😄 

Ici je vois que la méthode testée est `certificationPartnerRepository.buildCertificationPartnerClea`. Elle retourne manifestement un objet. Est-ce que quelque chose empêche de tester cette fonction en observant simplement sa valeur de retour (l'objet en question) ? Ça rendrait le test certes un peu moins "unitaire" et un plus "sociable" (le code du constructeur de `CertificationPartnerClea` étant impliqué) mais aussi nettement plus pertinent. Ici par exemple, je crois que le test passe toujours si on enlève le mot `return` de la fonction `buildCertificationPartnerClea` 🤷‍♂️

Reply:
J'aime bien ce petit talk sur les appels sortants à tester/ne pas tester : https://www.youtube.com/watch?v=URSWYvyc42M

Reply:
J'ai découvert il n'y a pas longtemps cette page de principes, et j'aime bien le principe "Use the Front Door First" :
- http://xunitpatterns.com/Principles%20of%20Test%20Automation.html#Use%20the%20Front%20Door%20First

Reply:
On a remplacé tout ca par un test de collaboration